File: api/Modules/Applicants/Entity/Applicant.cs

using api.Modules.Dependents.Entity;
using api.Modules.Hubs.Entity;

namespace api.Modules.Applicants.Entity;


public class Applicant
{
  public Guid Id { get; init; }
  public string Name { get; private set; }
  public string CPF { get; private set; }
  public string Email { get; private set; }
  public string PhoneNumber { get; private set; }
  public string Address { get; private set; }
  public bool IsBeneficiary { get; private set; }
  public int BeneficiaryQtd { get; private set; } = 0;
  public DateTime CreatedAt { get; private init; } = DateTime.UtcNow;
  public DateTime UpdatedAt { get; private set; } = DateTime.UtcNow;

  // URL da foto de perfil
  public string? ProfileImageUrl { get; private set; }

  // Relacionamento com Hub
  public Guid HubId { get; private set; }
  public Hub Hub { get; private set; } = null!;

  // Propriedade de navegação
  public ICollection<Dependent>? Dependents { get; init; } = [];

  public Applicant(string name, string cpf, string email, string phoneNumber, string address, bool isBeneficiary, Guid hubId, string? profileImageUrl = null)
  {
    Id = Guid.NewGuid();
    Name = name;
    CPF = cpf;
    Email = email;
    PhoneNumber = phoneNumber;
    Address = address;
    IsBeneficiary = isBeneficiary;
    HubId = hubId;
    ProfileImageUrl = profileImageUrl;
  }

  private Applicant()
  {
    Id = Guid.NewGuid();
    Name = string.Empty;
    CPF = string.Empty;
    Email = string.Empty;
    PhoneNumber = string.Empty;
    Address = string.Empty;
    IsBeneficiary = false;
    HubId = Guid.Empty;
    ProfileImageUrl = null;
  }
  public void SetProfileImageUrl(string? url)
  {
    ProfileImageUrl = url;
  }

  public void SetName(string name)
  {
    Name = name;
  }

  public void SetCPF(string cpf)
  {
    CPF = cpf;
  }

  public void SetEmail(string email)
  {
    Email = email;
  }

  public void SetPhoneNumber(string phoneNumber)
  {
    PhoneNumber = phoneNumber;
  }

  public void SetAddress(string address)
  {
    Address = address;
  }

  public void SetIsBeneficiary(bool isBeneficiary)
  {
    IsBeneficiary = isBeneficiary;
  }

  public void SetBeneficiaryQtd(int beneficiaryQtd)
  {
    BeneficiaryQtd = beneficiaryQtd;
  }

  public void UpdateTimestamps()
  {
    UpdatedAt = DateTime.UtcNow;
  }
}

File: project_rotary/lib/app/pdt/applicants/domain/entities/applicant.dart
class Applicant {
  final String id;
  final String name;
  final String cpf;
  final String email;
  final String phoneNumber;
  final String? address;
  final bool isBeneficiary;
  final List<String> beneficiaryIds;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Applicant({
    required this.id,
    required this.name,
    required this.cpf,
    required this.email,
    required this.phoneNumber,
    this.address,
    required this.isBeneficiary,
    required this.beneficiaryIds,
    required this.createdAt,
    this.updatedAt,
  });

  factory Applicant.fromJson(Map<String, dynamic> json) {
    return Applicant(
      id: json['id'] as String,
      name: json['name'] as String,
      cpf: json['cpf'] as String,
      email: json['email'] as String,
      phoneNumber: json['phoneNumber'] as String,
      address: json['address'] as String?,
      isBeneficiary: json['isBeneficiary'] as bool,
      beneficiaryIds: List<String>.from(json['beneficiaryIds'] ?? []),
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt:
          json['updatedAt'] != null
              ? DateTime.parse(json['updatedAt'] as String)
              : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'cpf': cpf,
      'email': email,
      'phoneNumber': phoneNumber,
      'address': address,
      'isBeneficiary': isBeneficiary,
      'beneficiaryIds': beneficiaryIds,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  Applicant copyWith({
    String? id,
    String? name,
    String? cpf,
    String? email,
    String? phoneNumber,
    String? address,
    bool? isBeneficiary,
    List<String>? beneficiaryIds,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Applicant(
      id: id ?? this.id,
      name: name ?? this.name,
      cpf: cpf ?? this.cpf,
      email: email ?? this.email,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      address: address ?? this.address,
      isBeneficiary: isBeneficiary ?? this.isBeneficiary,
      beneficiaryIds: beneficiaryIds ?? this.beneficiaryIds,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Applicant && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'Applicant(id: $id, name: $name, cpf: $cpf)';
  }
}
